statistical_disclosure_control/cell_key_perturbation/cell_key_perturbation_specification.md

```diff
@@ -0,0 +1,183 @@
+ |pcv        | integer || 0-750 | No intrinsic meaning. This value is an edited version of the count designed to reduce the necessary size of the ptable file. For cell values 0-750, pcv is the same as cell value | One value per cell in a frequency table | | |     
+ |count     | integer  || >=0  | The cell value after perturbation. The 'count' of how many records contain a combination of characteristics, plus the noise added by perturbation. E.g. there could be 14 males aged 18-24 in an area. Noise of +1 is applied to the cell so the post perturbation 'count' is 15. | One value per cell in a frequency table | | | 
+<!-- markdownlint-enable -->
+
+---
+
+* The count table will be in 'long' format sometimes known as tidy format, with
+one row per combination of characteristics. One column is included for each
+variable tabulated, e.g.: Local Authority, Age, Sex, count, pcv, pvalue, count;
+Bristol, 16-24, 1, 503, 503, +2, 505.
+
+## 5.3 Error handling  
+
+---
+
+Input data will need to be checked for errors, and that data contains record
+keys in the appropriate format.
+
+Any planned data editing should also be performed before the perturbation. The
+method will assume the data is correct and produce outputs accordingly.
+
+Where data contain missing values, the method will by default carry out
+perturbation and produce results for missing categories. If categories are to
+be combined or values imputed to replace missing data, this should be done
+before the perturbation.
+
+## 6.0 Overall Method  
+
+---
+
+The code produces a frequency table, counting how many records contain
+combinations of each of the supplied variables. In the same step, it takes the
+sum of the record keys of those records and takes modulo 256. The modulo is
+used so that cell keys are uniform in the range 0-255 (or 0-4095), which allows
+the method to specify and control the perturbation rate.
+
+To prevent the need for a very large ptable file, the rows 501-750 are also
+used for higher cell counts. Row 501 is used for cell counts 501, 751, 1001,
+1251... Row 502 is used for cell counts 502, 752, 1002. To achieve this effect,
+a column named \'pcv\' (perturbation cell value) is created and cell counts
+above 750 are transformed by subtracting 1, taking modulo 250 (to get the range
+0-249) then adding 501 (to get the range 501-750).
+
+The ptable file is merged on using a left join, so every cell in the table is
+covered, matching on values of cell-key and pcv. The post-perturbation count
+is created by adding the pre-perturbation count and the designated noise from
+the ptable, the 'pvalue' column (perturbation value).
+
+The end result is a table of combinations of variables, provided by the user,
+and the 'count' column has had the required noise from the ptable applied.
+
+## 7.0 Algorithm  
+
+---
+
+The user supplies the data to be analysed, variables to be tabulated including
```

Comment:
Other method specs outline the mathematical formulae for the steps of the algorithm that is being described? I am happy if you feel the information here is enough to make the method public.